compiler/dyno/lib/resolution/Resolver.cpp

```diff
@@ -745,28 +746,95 @@ void Resolver::resolveTupleDecl(const TupleDecl* td,
   resolveTupleUnpackDecl(td, useT);
 }
 
-bool Resolver::resolveSpecialCall(const Call* call) {
-  if (auto op = call->toOpCall()) {
-    if (op->op() == USTR("=")) {
-      if (op->numActuals() == 2) {
-        if (auto lhsTuple = op->actual(0)->toTuple()) {
-          QualifiedType lhsType = byPostorder.byAst(op->actual(0)).type();
-          QualifiedType rhsType = byPostorder.byAst(op->actual(1)).type();
-
-          resolveTupleUnpackAssign(lhsTuple, lhsType, rhsType);
-          return true;
-        }
+bool Resolver::resolveSpecialNewCall(const Call* call) {
+  if (!call->calledExpression() ||
+      !call->calledExpression()->isNew()) {
+    return false;
+  }
+
+  auto newExpr = call->calledExpression()->toNew();
+  ResolvedExpression& re = byPostorder.byAst(call);
+
+  // the type of the 'new' call is the type of the 'new' expression
+  ResolvedExpression& reNewExpr = byPostorder.byAst(newExpr);
+  re.setType(reNewExpr.type());
```

Comment:
I don't understand - isn't `init` a void method? I think you are correct that what needs to happen here is that we take the initial type from the `new R` and then use it + the actuals to compute the concrete type. Are you suggesting that I can do that quickly by just grabbing the receiver type out of the `init` call? I thought I might have to pre-compute the concrete receiver type so that I can pass it in to do the instantiation of `init`.

Reply:
> I don't understand - isn't init a void method?

Returns void, yes

>  I think you are correct that what needs to happen here is that we take the initial type from the new R and then use it + the actuals to compute the concrete type.

We are on the same page, then

> Are you suggesting that I can do that quickly by just grabbing the receiver type out of the init call?

No, I don't think that will help. I was just asking for a TODO comment.

> I thought I might have to pre-compute the concrete receiver type so that I can pass it in to do the instantiation of init.

You have to compute it while resolving the `init`

Reply:
Added a TODO to compute the concrete receiver type